src/java/org/apache/cassandra/db/SinglePartitionReadCommand.java

```diff
@@ -712,15 +713,20 @@ private UnfilteredRowIterator queryMemtableAndDiskInternal(ColumnFamilyStore cfs
         }
     }
 
-    private boolean shouldInclude(SSTableReader sstable)
+    private boolean intersects(SSTableReader sstable)
     {
+        return clusteringIndexFilter().intersects(sstable.metadata().comparator, sstable.getSSTableMetadata().coveredClustering);
+    }
+
+    private boolean hasRequiredStatics(SSTableReader sstable) {
         // If some static columns are queried, we should always include the sstable: the clustering values stats of the sstable
         // don't tell us if the sstable contains static values in particular.
-        // TODO: we could record if a sstable contains any static value at all.
-        if (!columnFilter().fetchedColumns().statics.isEmpty())
-            return true;
+        return !columnFilter().fetchedColumns().statics.isEmpty() && sstable.header.hasStatic();
+    }
 
-        return clusteringIndexFilter().intersects(sstable.metadata().comparator, sstable.getSSTableMetadata().coveredClustering);
+    private boolean hasPartitionLevelDeletions(SSTableReader sstable)
+    {
+        return sstable.getSSTableMetadata().hasPartitionLevelDeletions && sstable.mayHaveTombstones();
```

Comment:
These should be the other way around -- `hasPartitionLevelDeletions` implies `mayHaveTombstones` thus we can save some work by failing faster on the latter being false.

Reply:
indeed, thanks